### ISSUE TITLE
[$200] Interactor CLI prompt print problem

### ISSUE THREAD

**Issue by dkochmanski** [robert-strandh/McCLIM#116]
When we use CLI (not accepting-values form) the snippet attached below doesn't print prompt for the second argument of `bar` command but the default value of the previous one.

Steps to reproduce:
1. compile snippet and run `(bug-clim:gui-main)`
2. type `bar [Space] [Space] [Space] [Space]` in the CLI

Expected result in the interactor pane:
```
Command: bar (Prompt1[G2042]) (Prompt2["bah"]) (Prompt3[None])
Command:
```

Observed result:
```
Command: bar (Prompt1[G2042]) #:G2042 (Prompt3[None])
```

```common-lisp
(defpackage #:bug-clim
  (:use #:clim-lisp #:clim)
  (:export #:gui-main))

(in-package #:bug-clim)

(define-application-frame bug? () ()
  (:panes
   (display :application :display-time nil)
   (interactor :interactor))

  (:layouts (default (vertically ()
                       (3/5 display)
                       (2/5 interactor)))))

(define-bug?-command (bar :menu t :name "bar")
    ((a symbol :default (gensym) :prompt "Prompt1")
     (b string :default "bah" :prompt "Prompt2")
     (c (or string null) :default nil :prompt "Prompt3"))
  (format t "~s ~s ~s~%" a b c))

(defun gui-main ()
  (run-frame-top-level (make-application-frame 'bug?)))
```

**Comment by gabriel-laddel** [comment]
```
(defun accept-1 (stream type &key
		 (view (stream-default-view stream))
		 (default nil defaultp)
		 (default-type nil default-type-p)
		 provide-default
		 insert-default
		 (replace-input t)
		 history
		 active-p
		 prompt
		 prompt-mode
		 display-default
		 query-identifier
		 (activation-gestures nil activationsp)
		 (additional-activation-gestures nil additional-activations-p)
		 (delimiter-gestures nil delimitersp)
		 (additional-delimiter-gestures nil  additional-delimiters-p))
  (declare (ignore provide-default history active-p
		   prompt prompt-mode
		   display-default query-identifier))
  (when (and defaultp (not default-type-p))
    (error ":default specified without :default-type"))
  (when (and activationsp additional-activations-p)
    (error "only one of :activation-gestures or ~
            :additional-activation-gestures may be passed to accept."))
  (unless (or activationsp additional-activations-p *activation-gestures*)
    (setq activation-gestures *standard-activation-gestures*))
  (let ((sensitizer-object nil)
	(sensitizer-type nil))
    (with-input-editing
	(stream
	 :input-sensitizer #'(lambda (stream cont)
			       (with-output-as-presentation
				   (stream sensitizer-object sensitizer-type)
				 (funcall cont))))
      (with-input-position (stream)	; support for calls to replace-input
        (when (and insert-default
                   (not (stream-rescanning-p stream)))
          ;; Insert the default value to the input stream. It should
          ;; become fully keyboard-editable. We do not want to insert
          ;; the default if we're rescanning, only during initial
          ;; setup.          
          (presentation-replace-input stream default default-type view))
	(setf (values sensitizer-object sensitizer-type)
              (with-input-context (type)
		  (object object-type event options)
		(with-activation-gestures ((if additional-activations-p
					       additional-activation-gestures
					       activation-gestures)
					   :override activationsp)
		  (with-delimiter-gestures ((if additional-delimiters-p
						additional-delimiter-gestures
						delimiter-gestures)
					    :override delimitersp)
		    (let ((accept-results nil))
		      (handle-empty-input (stream)
                        (setq accept-results
                              (multiple-value-list
                               (if defaultp
                                   (funcall-presentation-generic-function
                                     accept type stream view
                                     :default default
                                     :default-type default-type)
                                   (funcall-presentation-generic-function
                                     accept type stream view))))
			;; User entered activation or delimiter
			;; gesture without any input.
			(if defaultp
                            (let* ((drei-view (drei::view (DREI:DREI-INSTANCE stream))))
                              (presentation-replace-input stream default default-type view :rescan nil)
                              (incf (stream-scan-pointer stream)))
			    (simple-parse-error "Empty input for type ~S with no supplied default" type))
			(setq accept-results (list default default-type)))
		      ;; Eat trailing activation gesture
		      ;; XXX what about pointer gestures?
		      ;; XXX and delimiter gestures?
		      (unless *recursive-accept-p*
			(let ((ag (read-char-no-hang stream nil stream t)))
			  (unless (or (null ag) (eq ag stream))
			    (unless (activation-gesture-p ag)
			      (unread-char ag stream)))))
		      (values (car accept-results) (if (cdr accept-results)
						       (cadr accept-results)
						       type)))))
		;; A presentation was clicked on, or something
                (t (when (and replace-input
                              (getf options :echo t)
                              (not (stream-rescanning-p stream)))
                     (presentation-replace-input stream object object-type view
                                                 :rescan nil))
                   (values object object-type))))
	;; Just to make it clear that we're returning values
	(values sensitizer-object sensitizer-type)))))
```

The problem is fixed by incrementing the STREAM-SCAN-POINTER. 

@dkochmanski lmk that this fix works for you, I'll make it a commit & do the songdance for the bounty.

**Comment by dkochmanski** [comment]
It doesn't solve my issue, what I have is:

`Command: bar (Prompt1[G1828]) #:G1828 (Prompt2[bah])`

**Comment by dkochmanski** [comment]
I've recompiled accept-1 in climi package with your code and compiled application from the issue description after that. Then I've called `gui-main`.

**Comment by gabriel-laddel** [comment]
@dkochmanski The string you wanted, "bar (Prompt1[G2042]) (Prompt2["bah"]) (Prompt3[None])" is wrong according to the spec (check the docs for ACCEPT-1). If you want the behavior anyways (ie, no PRESENTATION-REPLACE-INPUT) I have the hack somewhere...

**Comment by dkochmanski** [comment]
I'll reread the spec with this regard and let you know.

**Comment by dkochmanski** [comment]
As far as I understand, correct result would be
```
Command: bar (Prompt1[G2042]) #:G1828 (Prompt2["bah"]) "bah" (Prompt3[None]) None
Command:
```

And what we have after your fix is:
```
Command: bar (Prompt1[G2042]) #:G1828 (Prompt2["bah"]) (Prompt3[None]) None
Command:
```

Am I right? In that case your fix goes in right direction, but default value for Prompt2 still isn't printed, so we still have a bug in there.

**Comment by dkochmanski** [comment]
maybe increasing scan-pointer by two will solve the issue?

**Comment by gabriel-laddel** [comment]
@dkochmanski 

My apologies, your original interpretation was correct. INSERT-DEFAULT is NOT called by PROMPT-FOR-ACCEPT, which is the alternate ACCEPT path used in the CLI, or triggered by #\, at the Listener. The correct behavior is, as you stated,

```Command: bar (Prompt1[G2042]) (Prompt2["bah"]) (Prompt3[None])```

One can verify this by inserting a,

```(stumpwm::message (format nil "~S ~S ~S" defualtp default insert-default))```

```
(defun accept-1 (stream type &key
		 (view (stream-default-view stream))
		 (default nil defaultp)
		 (default-type nil default-type-p)
		 provide-default
		 insert-default
		 (replace-input t)
		 history
		 active-p
		 prompt
		 prompt-mode
		 display-default
		 query-identifier
		 (activation-gestures nil activationsp)
		 (additional-activation-gestures nil additional-activations-p)
		 (delimiter-gestures nil delimitersp)
		 (additional-delimiter-gestures nil  additional-delimiters-p))
  (declare (ignore provide-default history active-p
		   prompt prompt-mode
		   display-default query-identifier))
  (stumpwm::message (mm::sformat "DEFUALT: ~S DEFAULTP: ~S INSERT-DEFUALT: ~S"
                                 default defaultp insert-default))
  (when (and defaultp (not default-type-p))
    (error ":default specified without :default-type"))
  (when (and activationsp additional-activations-p)
    (error "only one of :activation-gestures or ~
            :additional-activation-gestures may be passed to accept."))
  (unless (or activationsp additional-activations-p *activation-gestures*)
    (setq activation-gestures *standard-activation-gestures*))
  (let ((sensitizer-object nil)
	(sensitizer-type nil))
    (with-input-editing
	(stream
	 :input-sensitizer #'(lambda (stream cont)
			       (with-output-as-presentation
				   (stream sensitizer-object sensitizer-type)
				 (funcall cont))))
      (with-input-position (stream)	; support for calls to replace-input
        (when (and insert-default
                   (not (stream-rescanning-p stream)))
          ;; Insert the default value to the input stream. It should
          ;; become fully keyboard-editable. We do not want to insert
          ;; the default if we're rescanning, only during initial
          ;; setup.          
          (presentation-replace-input stream default default-type view))
	(setf (values sensitizer-object sensitizer-type)
              (with-input-context (type)
		  (object object-type event options)
		(with-activation-gestures ((if additional-activations-p
					       additional-activation-gestures
					       activation-gestures)
					   :override activationsp)
		  (with-delimiter-gestures ((if additional-delimiters-p
						additional-delimiter-gestures
						delimiter-gestures)
					    :override delimitersp)
		    (let ((accept-results nil))
		      (handle-empty-input (stream)
                        (setq accept-results
                              (multiple-value-list
                               (if defaultp
                                   (funcall-presentation-generic-function
                                     accept type stream view
                                     :default default
                                     :default-type default-type)
                                   (funcall-presentation-generic-function
                                     accept type stream view))))
			;; User entered activation or delimiter
			;; gesture without any input.
                        (cond ((and defaultp insert-default)
                               (let* ((drei-view (drei::view (DREI:DREI-INSTANCE stream))))
                                 (presentation-replace-input stream default default-type view :rescan nil)
                                 (incf (stream-scan-pointer stream))))
                              (defaultp (incf (stream-scan-pointer stream)))
                              (t (simple-parse-error "Empty input for type ~S with no supplied default" type)))
			(setq accept-results (list default default-type)))
		      ;; Eat trailing activation gesture
		      ;; XXX what about pointer gestures?
		      ;; XXX and delimiter gestures?
		      (unless *recursive-accept-p*
			(let ((ag (read-char-no-hang stream nil stream t)))
			  (unless (or (null ag) (eq ag stream))
			    (unless (activation-gesture-p ag)
			      (unread-char ag stream)))))
		      (values (car accept-results) (if (cdr accept-results)
						       (cadr accept-results)
						       type)))))
		;; A presentation was clicked on, or something
                (t (when (and replace-input
                              (getf options :echo t)
                              (not (stream-rescanning-p stream)))
                     (presentation-replace-input stream object object-type view
                                                 :rescan nil))
                   (values object object-type))))
	;; Just to make it clear that we're returning values
	(values sensitizer-object sensitizer-type)))))
```

**Comment by gabriel-laddel** [comment]
The reason that the ACCEPT-1 for symbols with default was acting oddly in the first place is because of the way that %ACCEPT is defined for symbols, specifcally the use of ACCEPT-USING-READ & READ-TOKEN.

```
(define-presentation-method accept ((type symbol) stream (view textual-view)
				    &key)
  (accept-using-read stream type))

(defun accept-using-read (stream ptype &key ((:read-eval *read-eval*) nil))
  (let* ((token (read-token stream)))
    (let ((result (handler-case (read-from-string token)
		    (error (c)
		      (declare (ignore c))
		      (simple-parse-error "Error parsing ~S for presentation type ~S"
					  token
					  ptype)))))
      (if (presentation-typep result ptype)
	  (values result ptype)
	  (input-not-of-required-type result ptype)))))
```

IMHO, you should accept either one of the patches I posted above as a solution, ignore for a moment the slightly annoying extra space. If you do, I will happily share my code that uses the (setf (documentation clim:symbol 'function) "...") to add the spec to function / var docstrings when complete. Right now it generates a CLIM-spec-in-CLIM, but I have not quite yet finished nailing down the correct print routines. If you want this codebase to move forwards substantially, this will save a huge amount of time vs jumping between windows to see wtf behavior is supposed to occur. Lazily hacking is how things get done, and why emacs is better than all comers.

**Comment by gabriel-laddel** [comment]
ACCEPT-1, ACCEPT & ACCEPTING-VALUES all are incomplete. People are happy to fix, but not when it is a PITA to see wtf even is the correct behavior

**Comment by gabriel-laddel** [comment]
Oh, to be clear, I already wrote the code once that removes the extra space -- cannot remember exactly what it was offhand, but something to do with redefining the %ACCEPT method for symbols. The problem with this is that ACCEPT, ACCEPT-1, %ACCEPT & ACCEPTING-VALUES, AND RELATED FUNCTIONALITY in CLIM is so far from the spec, and in some cases just not implemented correctly that I hesitate to add "just another hack". What is NEEDED is for someone to stare down the logic of the whole presentation system and unify it. 

There are comments scattered throughout the codebase that indicate this, and broken behaviors too. For example, if you PRESENT an object on a textual view, hover your mouse over it, then remove, it will redraw itself... in reverse Z-ordering. Not all the time, but when layering multiple images & the like. Here is the fix:

```
(defmethod map-over-output-records-1
    ;; NOTE, @2017-02-15T03:45:46.009129Z
    ;; from /root/quicklisp/local-projects/McCLIM/Core/clim-basic/
    ;; without the 2x added REVERSE, the output record will be displayed in
    ;; reverse order
    (function (record standard-sequence-output-record) function-args)
  "Applies FUNCTION to all children in the order they were added."
  (if function-args
      (loop with children = (reverse (output-record-children record))
            for child across children
            do (apply function child function-args))
      (loop with children = (reverse (output-record-children record))
            for child across children
            do (funcall function child))))
```

To reiterate my point -- one-off hacks to fix issues that don't prevent the use of CLIM is silly. Having people GROK the spec is the correct direction.

**Comment by dkochmanski** [comment]
While I agree that reading the spec and fixing things at one go would be the best solution, given that nobody grokked it yet is a strong indicator, that we may wait for that another 17 years.

Fixing annoying bugs is the second best option, because we are getting closer to the correct solution one-by-one. Also it is more possible, that we'll gain more users because they won't be discouraged by annoying bugs scattered all over the place - some of them may become developers, and maybe one of them will actually grok the spec and fix the model (if indeed the model is broken).

As of the annoying space: I don't mind "having" it, but the behavior must be consistent between strings and symbols (in this case strings require single space, while other types require extra space) – not having extra space at all would be optimal of course.

If you are willing to fix the extra space problem that it is not annoying and works the same for any argument, I will raise the bounty to $200. Let me know if you are interested in that.

Regarding style, I know that this is only the solution draft (your code snippet), but for pull request I have a few remarks:
- Please put only spaces (don't introduce tabs where they were not present) – that makes diffs atracious,
- I've spotted, that you create a let for `drei-view`, but you don't use the mentioned `drei-view` at all – is it a typo?
- If you change indentation of parts of the code which are not revelant to the fix, put them in a separate commit (preferably with commit message `cosmetic: indent`.

For instance, your accept-1 change would be much easier for me to analyze, if it were just:
```diff
diff --git a/Core/clim-core/presentation-defs.lisp b/Core/clim-core/presentation-defs.lisp
index 4a2f74b..40eb041 100644
--- a/Core/clim-core/presentation-defs.lisp
+++ b/Core/clim-core/presentation-defs.lisp
@@ -874,13 +874,11 @@ history will be unchanged."
                                         accept type stream view))))
                           ;; User entered activation or delimiter
                           ;; gesture without any input.
-                          (if defaultp
-                              (progn
-                                (presentation-replace-input
-                                 stream default default-type view :rescan nil))
-                              (simple-parse-error
-                               "Empty input for type ~S with no supplied default"
-                               type))
+                          (cond ((and defaultp insert-default)
+                                 (presentation-replace-input stream default default-type view :rescan nil)
+                                 (incf (stream-scan-pointer stream)))
+                                (defaultp (incf (stream-scan-pointer stream)))
+                                (t (simple-parse-error "Empty input for type ~S with no supplied default" type)))
                           (setq accept-results (list default default-type)))
                         ;; Eat trailing activation gesture
                         ;; XXX what about pointer gestures?
```

**Comment by gabriel-laddel** [comment]
"If you are willing to fix the extra space problem that it is not annoying and works the same for any argument, I will raise the bounty to $200. Let me know if you are interested in that."

Very interested. Will clean up the commit.

**Comment by gabriel-laddel** [comment]
Finished. Le proof: 

![screenshot-3705448033](https://cloud.githubusercontent.com/assets/1878709/26750083/8b2358be-4809-11e7-9e7e-9626133e36cf.png)

**Comment by gabriel-laddel** [comment]
@dkochmanski If I post the COMPLETE FIX in a code-block here, are you going to grant me the bounty? 

I am happy to submit a patch, but need to merge with MASTER first (24 hours). I simply don't wish for anyone to take the code & make a commit -- and then have you grant them the bounty. Which tbh, idk what you would do in that situation. So please, do enlighten me, because I also solved the multilanguage bounty.

**Comment by dkochmanski** [comment]
I'm under control of bounties, please make a PR, I assure you that when it's merged you'll get the bounty. It's not FIFO, I know that you work on it.

If I'll have some remarks regarding PR, I may ask you to fix some things before merge and concluding the issue as resolved.

**Comment by gabriel-laddel** [comment]
Here is the fix. Tested it in all places WHO-CALLS shows + everything I could think of. Pull request impending.

```
(in-package climi)

(defun accept-using-read (stream ptype &key ((:read-eval *read-eval*) nil))
  (let* ((token (read-token stream)))
    (if (string= "" token)
        (values nil ptype)
        (let ((result (handler-case (read-from-string token)
                        (error (c)
                               (ignore c)
                               (simple-parse-error "Error parsing ~S for presentation type ~S"
                                                   token
                                                   ptype)))))
          (if (presentation-typep result ptype)
              (values result ptype)
              (input-not-of-required-type result ptype))))))

(define-presentation-method accept ((type symbol) stream (view textual-view)
                                                  &key (default-type type)
                                                  default)
  (let* ((read-result (accept-using-read stream type)))
    (if (and (null read-result) default) 
        (values default type)
        (values read-result type))))

(DEFINE-PRESENTATION-METHOD ACCEPT ((TYPE REAL) STREAM (VIEW TEXTUAL-VIEW) &KEY 
                                                (default-type type)
                                                default)
  (let* ((*READ-BASE* CLIM-INTERNALS::BASE)
         (read-result (accept-using-read stream type)))
    (if (and (null read-result) default) 
        (values default type)
        (values read-result type))))
```

**Comment by dkochmanski** [comment]
Case linked in the issue description works fine for me with your code. lmk when you issue a PR. Thanks!

**Comment by dkochmanski** [comment]
bump about the pull request :)

**Comment by dkochmanski** [comment]
Fixed by #266
